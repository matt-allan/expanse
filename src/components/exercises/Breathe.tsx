import React, { useEffect, useState } from "react";
import styled, { keyframes } from "styled-components";
import { defaultProps, Button } from "grommet";

// Only undefined if you overwrite the theme and don't deepMerge.
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const colors = defaultProps!.theme!.global!.colors!;

const pulse = keyframes`
  0% {
    transform: scale(0.1) rotate(180deg);
  }
  100% {
    transform: scale(0.1);
  }
}
`;

const resize = keyframes`
  0% {
    transform: scale(1) 
  }
  100% {
    transform: scale(10) 
  }
`;

const Container = styled.div`
  height: 125px;
  width: 125px;
  animation: ${pulse} 4s ease alternate infinite;
`;

const Circle = styled.div`
  height: 125px;
  width: 125px;
  border-radius: 50%;
  position: absolute;
  top: 50px;
  left: 50px;
  opacity: 0.75;
  animation: ${resize} 4s ease alternate infinite;
  mix-blend-mode: screen;

  &:nth-child(odd) {
    background: ${colors["accent-1"]};
  }

  &:nth-child(even) {
    background: ${colors["accent-3"]};
  }

  &:nth-child(1) {
    transform-origin: 95% 75%;
  }

  &:nth-child(2) {
    transform-origin: 50% 100%;
  }

  &:nth-child(3) {
    transform-origin: 5% 75%;
  }

  &:nth-child(4) {
    transform-origin: 95% 25%;
  }

  &:nth-child(5) {
    transform-origin: 50% 0%;
  }

  &:nth-child(6) {
    transform-origin: 5% 25%;
  }
`;

type BreatheProps = {
  onEnd: () => void;
};

export const Breathe = ({ onEnd }: BreatheProps): JSX.Element => {
  const [finished, setFinished] = useState(false);

  useEffect(() => {
    const timer = setTimeout(() => setFinished(true), 1000 * 60 * 2);

    return () => clearTimeout(timer);
  }, []);

  if (finished) {
    return <Button primary label="Back to work" onClick={onEnd} />;
  }

  return (
    <Container>
      <Circle />
      <Circle />
      <Circle />
      <Circle />
      <Circle />
      <Circle />
    </Container>
  );
};
