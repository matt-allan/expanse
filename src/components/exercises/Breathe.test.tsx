import React from "react";
import { shallow } from "enzyme";

import { Breathe } from "./Breathe";

jest.useFakeTimers();

it("renders", () => {
  const component = shallow(<Breathe onEnd={jest.fn()} />);

  expect(component).toMatchSnapshot();
});
