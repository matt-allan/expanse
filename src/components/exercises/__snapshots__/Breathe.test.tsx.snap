// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Component>
  <styled.div>
    <styled.div />
    <styled.div />
    <styled.div />
    <styled.div />
    <styled.div />
    <styled.div />
  </styled.div>
  <Component
    frames={
      Array [
        Object {
          "caption": "Bring your attention to your breath.",
          "end": 7,
          "start": 0,
        },
        Object {
          "caption": "Now inhale...",
          "end": 12,
          "start": 8,
        },
        Object {
          "caption": "and exhale.",
          "end": 15,
          "start": 13,
        },
      ]
    }
  />
</Component>
`;
